compiler/src/iree/compiler/Codegen/Common/ForallToFor.cpp

```diff
@@ -0,0 +1,119 @@
+  auto buildBody = [&](OpBuilder &builder, Location loc, ValueRange ivs,
+                       ValueRange args) -> scf::ValueVector {
+    // Inline `scf.forall` body excluding terminating
+    // `scf.forall.in_parallel` op.
+    IRMapping map;
+    map.map(forallOp.getInductionVars(), ivs);
+    map.map(forallOp.getRegionOutArgs(), args);
+    for (auto &op : forallOp.getBody()->without_terminator()) {
+      builder.clone(op, map);
+    }
+
+    // Convert + inline the contents of `scf.forall.in_parallel` terminator.
+    auto terminator = forallOp.getTerminator();
+    SmallVector<Value> yieldedValues;
+    for (auto &yieldOp : terminator.getYieldingOps()) {
+      // Convert tensor.parallel_insert_slice to tensor.insert_slice
+      if (auto parallelInsert =
+              dyn_cast<tensor::ParallelInsertSliceOp>(&yieldOp)) {
+
+        auto source = map.lookupOrDefault(parallelInsert.getSource());
+        auto dest = map.lookupOrDefault(parallelInsert.getDest());
+        auto offsets =
+            llvm::map_to_vector(parallelInsert.getOffsets(), [&](Value v) {
+              return map.lookupOrDefault(v);
+            });
+        auto sizes =
+            llvm::map_to_vector(parallelInsert.getSizes(), [&](Value v) {
+              return map.lookupOrDefault(v);
+            });
+        auto strides =
+            llvm::map_to_vector(parallelInsert.getStrides(), [&](Value v) {
+              return map.lookupOrDefault(v);
+            });
+        auto insertSlice = builder.create<tensor::InsertSliceOp>(
+            parallelInsert.getLoc(), source, dest, offsets, sizes, strides,
+            parallelInsert.getStaticOffsets(), parallelInsert.getStaticSizes(),
+            parallelInsert.getStaticStrides());
+        yieldedValues.push_back(insertSlice.getResult());
+      } else {
+        forallOp.emitError("unsupported operation in in_parallel region");
+        unsupportedOperation = true;
+        return args;
+      }
+    }
+    return yieldedValues;
+  };
+  scf::LoopNest loopNest = scf::buildLoopNest(rewriter, forallOp->getLoc(), lbs,
+                                              ubs, steps, iterArgs, buildBody);
+  if (unsupportedOperation) {
+    return failure();
+  }
+  rewriter.replaceOp(forallOp, loopNest.results);
+  return success();
+}
+
+struct ForallToForPass : impl::ForallToForPassBase<ForallToForPass> {
+  using impl::ForallToForPassBase<ForallToForPass>::ForallToForPassBase;
+  void runOnOperation() override {
+    auto funcOp = getOperation();
+    IRRewriter rewriter(funcOp->getContext());
+
+    // Find `scf.forall` ops we want to convert in innermost to outermost order.
+    SmallVector<scf::ForallOp> forallOps;
+    funcOp->walk<WalkOrder::PostOrder>([&](scf::ForallOp forallOp) {
+      // Forall ops with workgroup mappings `#iree_codegen.workgroup_mapping<y>`
+      // are for workgroup distribution, we only want to convert inner loops
+      // produced by tiling to `scf.for`.
```

Comment:
I think we can make the comment a little more accurate. It is not IREE specific, but more like distribution in general.

```suggestion
      // Forall ops with workgroup any mapping
      // are for distribution, we only want to convert inner loops
      // produced by tiling to `scf.for`.
```

Reply:
Can you format the comment a bit? I think the first line does not reach 80 chars. I edited it on Github UI, so I did not format it well. :)

Reply:
Yes, I've got some formatting in an upcoming commit, just haven't pushed yet :) resolving the other comments first.